onmt/translate/translation_server.py

```diff
@@ -518,43 +530,43 @@ def run(self, inputs):
                 tok = self.maybe_tokenize(seg)
                 if ref is not None:
                     ref = self.maybe_tokenize(ref, side='tgt')
-                inferred_feats = self.transform_feats(seg, tok, feats)
-                texts.append((tok, ref, inferred_feats))
+                feats = self.maybe_transform_feats(seg, tok, feats)
+                texts.append((tok, ref, feats))
             tail_spaces.append(whitespaces_after)
 
         empty_indices = []
-        texts_to_translate, texts_ref = [], []
-        texts_features = defaultdict(list)
+        examples = []
         for i, (tok, ref_tok, feats) in enumerate(texts):
             if tok == "":
                 empty_indices.append(i)
             else:
-                texts_to_translate.append(tok)
-                texts_ref.append(ref_tok)
-                for feat_name, feat_values in feats.items():
-                    texts_features[feat_name].append(feat_values)
-        if any([item is None for item in texts_ref]):
-            texts_ref = None
+                ex = {
+                    "src": {"src": tok},
+                    "tgt": {"tgt": ref_tok} if ref_tok is not None else None
+                }
+                if feats is not None:
+                    ex["src"]["feats"] = feats
+                examples.append(ex)
 
         scores = []
         predictions = []
-
-        if len(texts_to_translate) > 0:
+        if len(examples) > 0:
             try:
                 if isinstance(self.translator, CTranslate2Translator):
+                    # TODO
```

Comment:
What's the main blocker here? If we don't want to support it, we might want to make it more explicit (some explicit warning/assert / mention somewhere in the docs?)

Reply:
As far as I recall source features could be passed to Ctranslate2. However, there is a line which says the opposite:
https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/translate/translation_server.py#L146

I guess we just need to remove that assertion and adapt it to the current workflow. I think Ctranslate2 changed `translate()` to `translate_batch()` too

Reply:
I confirm that CTranslate2 supports source features: https://opennmt.net/CTranslate2/translation.html#source-factors